File: src/main/java/com/example/indentory_management_system/Service/StockService.java
package com.example.indentory_management_system.Service;

import java.util.List;

import com.example.indentory_management_system.dto.StockRequestdto;
import com.example.indentory_management_system.dto.StockResponsedto;

public interface StockService {

    StockResponsedto addStock(StockRequestdto dto);

    List<StockResponsedto> getAllStocks();

    // StockResponsedto stockOut(Long id, StockRequestdto dto);

    StockResponsedto updateStock(Long id, StockRequestdto dto);

    void deleteStock(Long id);

    // StockResponsedto adjustStock(Long id, int quantity);

    List<StockResponsedto> transferStock(Long fromWarehouseId, Long toWarehouseId, StockRequestdto dto);

    StockResponsedto getCurrentStock(Long productId, Long warehouseId);

    // List<StockResponsedto> getStockByProduct(String productName);

    // List<StockResponsedto> getStockByWarehouse(String warehouseName);
}

File: IMS_frontend/IMS/src/app/core/services/stock.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { StockRequestDto, StockResponseDto, StockTransferRequestDto } from '../../shared/models/interfaces';

@Injectable({
  providedIn: 'root'
})
export class StockService {
  private readonly apiUrl = 'http://localhost:8082/api/v1/stock';

  constructor(private http: HttpClient) {}

  addStock(dto: StockRequestDto): Observable<StockResponseDto> {
    return this.http.post<StockResponseDto>(`${this.apiUrl}/addstock`, dto);
  }

  getAllStocks(): Observable<StockResponseDto[]> {
    return this.http.get<StockResponseDto[]>(`${this.apiUrl}/getallstock`);
  }

  updateStock(id: number, dto: StockRequestDto): Observable<StockResponseDto> {
    return this.http.put<StockResponseDto>(`${this.apiUrl}/updatestock/${id}`, dto);
  }

  deleteStock(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/deletestock/${id}`);
  }

  transferStock(dto: StockTransferRequestDto): Observable<StockResponseDto[]> {
    return this.http.post<StockResponseDto[]>(`${this.apiUrl}/transferstock`, dto);
  }

  getCurrentStock(productId: number, warehouseId: number): Observable<StockResponseDto> {
    return this.http.get<StockResponseDto>(`${this.apiUrl}/getcurrentstock/${productId}/${warehouseId}`);
  }
}
